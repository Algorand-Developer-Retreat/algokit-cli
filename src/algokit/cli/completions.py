import logging
from pathlib import Path

import click
import click.shell_completion
import shellingham  # type: ignore
from algokit.core.atomic_write import atomic_write
from algokit.core.conf import get_app_config_dir

logger = logging.getLogger(__name__)

SUPPORTED_SHELLS = ["bash", "zsh"]


@click.group("completions", short_help="Install and Uninstall AlgoKit shell integration.")
def completions_group() -> None:
    pass


shell_option = click.option(
    "--shell", type=click.Choice(SUPPORTED_SHELLS), help="Specify shell to install algokit completions for."
)


@completions_group.command(
    "install",
    short_help="Install shell completions, this command will attempt to update the interactive profile script"
    " for the current shell to support algokit completions. To specify a specific shell use --shell.",
)
@shell_option
def install(shell: str | None) -> None:
    shell_completion = ShellCompletion(shell)
    shell_completion.install()


@completions_group.command(
    "uninstall",
    short_help="Uninstall shell completions, this command will attempt to update the interactive profile script"
    " for the current shell to remove any algokit completions that have been added."
    " To specify a specific shell use --shell.",
)
@shell_option
def uninstall(shell: str | None) -> None:
    shell_completion = ShellCompletion(shell)
    shell_completion.uninstall()


class ShellCompletion:
    def __init__(self, shell: str | None) -> None:
        shell = shell or _get_current_shell()
        self.shell = shell
        self.source_path = get_app_config_dir() / f".algokit-completions.{shell}"
        self.profile_path = Path(f"~/.{shell}rc").expanduser()
        home_based_source_path = _get_home_based_path(self.source_path)
        self.profile_line = f". {home_based_source_path}\n"

    def install(self) -> None:
        self._save_source()
        self._insert_profile_line()
        logger.info(f"AlgoKit completions installed for {self.shell} ðŸŽ‰")
        home_based_profile_path = _get_home_based_path(self.profile_path)
        logger.info(f"Restart shell or run `. {home_based_profile_path}` to enable completions")

    def uninstall(self) -> None:
        self._remove_source()
        self._remove_profile_line()
        logger.info(f"AlgoKit completions uninstalled for {self.shell} ðŸŽ‰")

    @property
    def source(self) -> str:
        completion_class = click.shell_completion.get_completion_class(self.shell)
        completion = completion_class(
            # class is only instantiated to get source snippet, so don't need to pass a real command
            None,  # type: ignore
            {},
            "algokit",
            "_ALGOKIT_COMPLETE",
        )
        try:
            return completion.source()
        except RuntimeError as ex:
            logger.debug(f"Failed to generate completion source. {ex}")
            if self.shell == "bash":
                logger.error("Shell completion is not supported for Bash versions older than 4.4.")
            else:
                logger.error("Failed to install completions ðŸ˜¢.")
            raise click.exceptions.Exit(code=1) from ex

    def _save_source(self) -> None:
        # grab source before attempting to write file in case it fails
        source = self.source
        logger.debug(f"Writing source script {self.source_path}")
        with open(self.source_path, "w") as source_file:
            source_file.write(source)
            source_file.flush()

    def _remove_source(self) -> None:
        logger.debug(f"Removing source script {self.source_path}")
        self.source_path.unlink(missing_ok=True)

    def _insert_profile_line(self) -> None:
        do_write = True
        if self.profile_path.exists():
            with open(self.profile_path) as file:
                for line in file:
                    if self.profile_line in line:
                        logger.debug(f"{self.profile_path} already contains completion source")
                        # profile already contains source of completion script. nothing to do
                        do_write = False
                        break

        if do_write:
            logger.debug(f"Appending completion source to {self.profile_path}")
            # got to end of file, so append profile line
            atomic_write([self.profile_line], self.profile_path, "a")

    def _remove_profile_line(self) -> None:
        if not self.profile_path.exists():
            logger.debug(f"{self.profile_path} not found")
            # nothing to do
            return

        # see if profile script contains profile_line, if it does remove it
        do_write = False
        lines = []
        with open(self.profile_path) as file:
            for line in file:
                if self.profile_line in line:
                    do_write = True
                    logger.debug(f"Completion source found in {self.profile_path}")
                else:
                    lines.append(line)

        if do_write:
            logger.debug(f"Removing completion source found in {self.profile_path}")
            atomic_write(lines, self.profile_path, "w")


def _get_home_based_path(path: Path) -> Path:
    home = Path("~").expanduser()
    try:
        home_based_path = path.relative_to(home)
        return "~" / home_based_path
    except ValueError:
        return path


def _get_current_shell() -> str:
    try:
        shell = shellingham.detect_shell()
        shell_name: str = shell[0]
    except Exception as ex:
        logger.debug("Could not determine current shell", exc_info=ex)
        logger.warning("Could not determine current shell. Try specifying a supported shell with --shell")
        raise click.exceptions.Exit(code=1) from ex

    if shell_name not in SUPPORTED_SHELLS:
        logger.warning(f"{shell_name} is not a supported shell. ðŸ˜¢")
        raise click.exceptions.Exit(code=1)
    return shell_name
