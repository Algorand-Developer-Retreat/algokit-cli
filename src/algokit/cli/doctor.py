import logging
import platform

import click
import pyperclip3
from algokit.core.doctor import DoctorFunctions

logger = logging.getLogger(__name__)
DOCTOR_END_MESSAGE = (
    "If you are experiencing a problem with algokit, feel free to submit an issue "
    "via https://github.com/algorandfoundation/algokit-cli/issues/new; please include this output, "
    "if you want to populate this message in your clipboard, run `algokit doctor -c`"
)


@click.command(
    "doctor",
    short_help="Run the Algorand doctor CLI.",
    context_settings={
        "ignore_unknown_options": True,
    },
)
@click.option(
    "--copy-to-clipboard",
    "-c",
    help="Copy the contents of the doctor message (in Markdown format) in your clipboard.",
    is_flag=True,
    default=False,
)
def doctor_command(*, copy_to_clipboard: bool) -> None:
    os_type = platform.system().lower()

    service_outputs: dict[str, str] = {}
    service_outputs_contents_lines: list[str] = []
    doctor_functions = DoctorFunctions()

    service_outputs["Time"] = doctor_functions.get_date()
    service_outputs["AlgoKit"] = doctor_functions.get_algokit_info()
    if os_type == "windows":
        service_outputs["Chocolatey"] = doctor_functions.get_choco_info()
    if os_type == "darwin":
        service_outputs["Brew"] = doctor_functions.get_brew_info()
    service_outputs["OS"] = doctor_functions.get_os(os_type)
    service_outputs["Docker"] = doctor_functions.get_docker_info()
    service_outputs["Docker Compose"] = doctor_functions.get_docker_compose_info()
    service_outputs["Git"] = doctor_functions.get_git_info(os_type)
    service_outputs["AlgoKit Python"] = doctor_functions.get_algokit_python_info()
    service_outputs["Global Python"] = doctor_functions.get_global_python_info()
    service_outputs["Pipx"] = doctor_functions.get_pipx_info()
    service_outputs["Poetry"] = doctor_functions.get_poetry_info()
    service_outputs["Node.js"] = doctor_functions.get_node_info()
    service_outputs["Npm"] = doctor_functions.get_npm_info()

    # Print the status details
    if copy_to_clipboard:
        for key, value in service_outputs.items():
            logger.info(click.style(f"{key}: ", bold=True) + f"{value}")
            service_outputs_contents_lines.append(f"{key}: {value}\n")
        service_outputs_contents_lines.append(DOCTOR_END_MESSAGE)
    else:
        for key, value in service_outputs.items():
            logger.info(click.style(f"{key}: ", bold=True) + f"{value}")

    # print end message anyway
    logger.info(DOCTOR_END_MESSAGE)

    if copy_to_clipboard:
        pyperclip3.copy("".join(service_outputs_contents_lines))
