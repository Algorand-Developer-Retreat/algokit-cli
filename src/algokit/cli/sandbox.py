import logging

import subprocess

import click

logger = logging.getLogger(__name__)

def run(command: str):
    return subprocess.run(
        command.split(),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )


@click.group("sandbox", short_help="Manage the Algorand sandbox")
def sandbox_group():
    logger.debug("Hello I'm the sandbox command group")


@sandbox_group.command("start")
def start_sandbox():
    print("Starting the AlgoKit sandbox now...")
    result = run("docker network create -d bridge algokit-network")
    click.echo(result.args)
    result = run(
        "docker run -p 4001:4001 -p 4002:4002 -p 9392:9392 "
        + '--network algokit-network --label com.docker.compose.project="AlgoKit_sandbox" '
        + "--name algokit_algod -d makerxau/algorand-sandbox-dev:latest"
    )
    click.echo(result.args)
    result = run(
        "docker run -u postgres -e POSTGRES_USER=algorand -e POSTGRES_PASSWORD=algorand -e POSTGRES_DB=indexer_db "
        + '--network algokit-network --label com.docker.compose.project="AlgoKit_sandbox" '
        + "--name algokit_postgres -d postgres:13-alpine"
    )
    click.echo(result.args)
    result = run(
        "docker run -p 8980:8980 -e ALGOD_HOST=algokit_algod -e POSTGRES_HOST=algokit_postgres -e POSTGRES_PORT=5432 "
        + "-e POSTGRES_USER=algorand -e POSTGRES_PASSWORD=algorand -e POSTGRES_DB=indexer_db "
        + '--network algokit-network --label com.docker.compose.project="AlgoKit_sandbox" '
        + "--name algokit_indexer -d makerxau/algorand-indexer-dev:latest "
    )
    click.echo(result.args)


@sandbox_group.command("restart")
def restart_sandbox():
    logger.info("Restarting the sandbox now...")
    # TODO: the thing
    logger.info("Done!")
