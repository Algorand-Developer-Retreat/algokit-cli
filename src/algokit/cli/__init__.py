import logging

import click

from algokit.core.conf import PACKAGE_NAME
from algokit.core.log_handlers import console_log_handler
from algokit.cli.init import init_command
from algokit.cli.sandbox import sandbox_group


@click.group(
    help="AlgoKit is your one-stop shop to develop applications on the Algorand blockchain.",
    context_settings={"help_option_names": ["-h", "--help"]},
)
@click.version_option(package_name=PACKAGE_NAME)
@click.option("--verbose", "-v", is_flag=True)
@click.option("--color/--no-color", default=None)
def algokit(verbose: bool, color: bool | None):
    if verbose:
        console_log_handler.setLevel(logging.DEBUG)
    if color is not None:
        console_log_handler.force_styles_to = color


algokit.add_command(init_command)
algokit.add_command(sandbox_group)
