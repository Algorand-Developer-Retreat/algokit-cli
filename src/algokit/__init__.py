import platform
import sys

import click

# this isn't beautiful, but to avoid confusing user errors we need this check before we start importing our own modules
if sys.version_info < (3, 10, 0):
    click.echo(
        click.style(
            f"Unsupported CPython version: {platform.python_version()} detected.\n"
            "The minimum version of Python supported is CPython 3.10.\n"
            "If you need help installing then this is a good starting point: \n"
            "https://www.python.org/about/gettingstarted/",
            fg="red",
        ),
        err=True,
    )
    sys.exit(-1)


if __name__ == "__main__":
    from algokit.cli import algokit

    algokit()
