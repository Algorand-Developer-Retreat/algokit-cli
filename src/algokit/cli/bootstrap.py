import logging
import sys

import click
from algokit.core import proc

logger = logging.getLogger(__name__)


@click.group("bootstrap", short_help="Bootstrap AlgoKit project dependencies.")
def bootstrap_group() -> None:
    pass


@bootstrap_group.command("poetry", short_help="Bootstrap Python Poetry and install in the current working directory.")
@click.option(
    "--ok-exit-code",
    default=False,
    help="Always return a 0 exit code; useful when calling as task from a Copier template to avoid template delete.",
    is_flag=True,
)
def poetry(*, ok_exit_code: bool) -> None:
    try:
        python = sys.executable
        try:
            proc.run(["poetry", "-V"])
        except IOError:
            # An IOError (such as PermissionError or FileNotFoundError) will only occur if "poetry"
            # isn't an executable in the user's path, which means poetry isn't installed
            try:
                proc.run(["pipx", "--version"])
            except IOError:
                # An IOError (such as PermissionError or FileNotFoundError) will only occur if "pipx"
                # isn't an executable in the user's path, which means pipx isn't installed

                proc.run(
                    [python, "-m", "pip", "install", "--user", "pipx"],
                    bad_return_code_error_message="Unable to install pipx; please install"
                    + " pipx manually via https://pypa.github.io/pipx/ and try `algokit bootstrap poetry` again.",
                )

                proc.run(
                    [python, "-m", "pipx", "ensurepath"],
                    bad_return_code_error_message="Unable to update pipx install path to global path;"
                    + " please set the path manually and try `algokit bootstrap poetry` again.",
                )

            proc.run(
                ["pipx", "install", "poetry"],
                bad_return_code_error_message="Unable to install poetry via pipx; please install poetry"
                + " manually via https://python-poetry.org/docs/ and try `algokit bootstrap poetry` again.",
            )

        logger.info("Installing Python dependencies and setting up Python virtual environment via Poetry")
        proc.run(["poetry", "install"], stdout_log_level=logging.INFO)

    except Exception as ex:
        logger.error(ex)
        if ok_exit_code:
            logger.error("Error bootstrapping poetry; try running `poetry install` manually.")
        else:
            raise click.ClickException("Error bootstrapping poetry; try running `poetry install` manually.") from ex
