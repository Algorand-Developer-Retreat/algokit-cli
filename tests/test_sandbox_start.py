from approvaltests import verify
from click.testing import CliRunner

from algokit.cli import algokit


def test_sandbox_start():
    runner = CliRunner()
    result = runner.invoke(algokit, ["sandbox", "start"])
    assert result.exit_code == 0
    verify(result.stdout)
