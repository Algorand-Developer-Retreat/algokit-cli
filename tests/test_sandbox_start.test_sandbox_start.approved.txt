Hello I'm the sandbox command group
Starting the AlgoKit sandbox now...
['docker', 'network', 'create', '-d', 'bridge', 'algokit-network']
['docker', 'run', '-p', '4001:4001', '-p', '4002:4002', '-p', '9392:9392', '--network', 'algokit-network', '--label', 'com.docker.compose.project="AlgoKit_sandbox"', '--name', 'algokit_algod', '-d', 'makerxau/algorand-sandbox-dev:latest']
['docker', 'run', '-u', 'postgres', '-e', 'POSTGRES_USER=algorand', '-e', 'POSTGRES_PASSWORD=algorand', '-e', 'POSTGRES_DB=indexer_db', '--network', 'algokit-network', '--label', 'com.docker.compose.project="AlgoKit_sandbox"', '--name', 'algokit_postgres', '-d', 'postgres:13-alpine']
['docker', 'run', '-p', '8980:8980', '-e', 'ALGOD_HOST=algokit_algod', '-e', 'POSTGRES_HOST=algokit_postgres', '-e', 'POSTGRES_PORT=5432', '-e', 'POSTGRES_USER=algorand', '-e', 'POSTGRES_PASSWORD=algorand', '-e', 'POSTGRES_DB=indexer_db', '--network', 'algokit-network', '--label', 'com.docker.compose.project="AlgoKit_sandbox"', '--name', 'algokit_indexer', '-d', 'makerxau/algorand-indexer-dev:latest']
